docs/networking.md

```diff
@@ -0,0 +1,114 @@
+2. A container is created using `web`'s configuration. It joins the network `myapp` under the name `web`.
+3. A container is created using `db`'s configuration. It joins the network `myapp` under the name `db`.
+
+Each container can now look up the hostname `web` or `db` and get back the appropriate container's IP address. For example, `web`'s application code could connect to the URL `postgres://db:5432` and start using the Postgres database.
+
+Because `web` explicitly maps a port, it's also accessible from the outside world via port 8000 on your Docker host's network interface.
+
+## Updating containers
+
+If you make a configuration change to a service and run `docker-compose up` to update it, the old container will be removed and the new one will join the network under a different IP address but the same name. Running containers will be able to look up that name and connect to the new address, but the old address will stop working.
+
+If any containers have connections open to the old container, they will be closed. It is a container's responsibility to detect this condition, look up the name again and reconnect.
+
+## Scaling services
+
+If you create multiple containers for a service with `docker-compose scale`, each container will join the network with the same name. For example, if you run `docker-compose scale web=3`, then 3 containers will join the network under the name `web`. Looking up the name `web` will return the IP address of one of them, but Docker and Compose do not provide any guarantees about which one.
+
+## Configure how services are published
+
+By default, containers for each service are published on the network with the same name as the service. If you want to change the name, or stop containers from being discoverable at all, you can use the `publish` option.
+
+`publish` takes a list of names. Each container for the service will join the network under each name.
+
+    web:
+      build: .
+      publish:
+        - "my-web-application"
+        - "other-name"
+
+If you specify `publish`, then the default behaviour of joining under the service name is overridden. If you still want to publish under the service name, add it back:
+
+    web:
+      build: .
+      publish:
+        - "my-web-application"
+        - "other-name"
+        - "web"
+
+You can stop a service's containers from being published at all by passing in an empty list.
+
+    web:
+      build: .
+      publish: []
+
+## Configure your own networks
+
+If the default behaviour of a single app-wide network isn't flexible enough, you can configure your own networks and specify what networks each container should join.
+
+Instead of defining your services at the top-level in `docker-compose.yml`, place them under a `services` key, and define networks under a top-level `networks` key.
+
+Here's how you would set up a 2-network, 3-container app where the `nginx` and `db` services cannot discover or communicate with each other directly:
+
```

Comment:
I think this section should focus on providing the flexibility to run multiple apps in a single network.  For example, `docker compose up -d --net=<network>` (and any equivalents of that which might feature directly within the Compose file if we want to support specifying networks in the Compose file).

I don't think support for containers connecting to multiple networks in 1.8 timeframe is realistic.  I think there are broad range of libnetwork, network driver and service discovery interaction and implementation issues that need to be resolved to make that work sensibly.  I suggest restricting each container to a single network for 1.8.